README.md
# hospital_clojure
 - Schemas

## Curso SCHEMA Clojure - Alura
- https://github.com/plumatic/schema


### Aula 01
O que aprendemos nesta aula:

- Usar schemas;
- Utilizar a biblioteca de schema plumatic;
- Como colocar dependência em um projeto;
- Validar os valores com base no schema;
- O que é o s/def;
- Verificar o tipo que vamos receber.

### Aula 02

O que aprendemos nesta aula:

- O que é s/defrecord é expansível e imutável;
- O que é explain;
- Criar schemas para números;
- O que é s/pred;
- Desativar ou ativar validação;
- Criar restrições;
- Utilizar o pos;

### Aula 03

O que aprendemos nesta aula:

- Utilizar uma função para validar os números positivos;
- O que são schemas compostos;
- Fazer a validação em sequência;
- Ter validação com valor obrigatório ou não;

### Aula 04

O que aprendemos nesta aula:

- Deixar uma chave opcional dentro de um schema;
- Utilizar mapa para ter as chaves dinâmicas.


### Aula 05

O que aprendemos nesta aula:

- Utilizar o modelo de schema;
- Garantir que nosso o schema está validando nosso modelo;

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-11-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-11-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hospital_clojure/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hospital_clojure/compare/0.1.0...0.1.1


test/hospital_clojure/core_test.clj
(ns hospital-clojure.core-test
  (:require [clojure.test :refer :all]
            [hospital-clojure.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/hospital_clojure/aula2.clj
(ns hospital-clojure.aula2
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

;;(s/defrecord Paciente
;  [id :- Long, nome :- s/Str])
;(pprint (Paciente. 15 "Amanda"))
;(pprint (Paciente. " 15" "Amanda"))
;(pprint (map->Paciente  {15 "Amanda"})
;(pprint (map->Paciente  {"15" "Amanda"})

(def Paciente
  "Schema de um paciente"
  {:id s/Num, :nome s/Str})

(pprint (s/explain Paciente))
(pprint (s/validate Paciente {:id 15, :nome "guilherme"}))

; tipo é pego pelo schema. mas poderiamos argumentar que esse
; tipo de erro seria pego em testes automatizados com cobertura boa
; (pprint (s/validate Paciente {:id 15, :nome "guilherme"}))

; mas... entra a questão do querer ser forward compatible OU NÃO
; entender esse trade-off
; sistemas externos não me quebrarão ao adicionar campos novos (foward compatible)
; no nosso validate não estamos sendo forward compatible (pode ser interessante quando quero analisar mudanças
;(pprint (s/validate Paciente {:id 15, :nome "guilherme", :plano [:raio-x]}))

; chaves que são keywords em schemas são por padrão OBRIGATORIAS
;(pprint (s/validate Paciente {:id 15})

; tipo de retorno com schema
; força a validação na saída da função
; (s/def novo-paciente :- Paciente
;   [id :- s/Num, nome :- s/Str]
;   {:id id, :nome nome, :plano []})

(s/defn novo-paciente :- Paciente
  [id :- s/Num, nome :- s/Str]
  {:id id, :nome nome})

(pprint (novo-paciente 15 "Guilherme"))

;Função pura simples, facil de testar
(defn estritamente-positivo? [x]
  (> x 0))

;(def EstritamentePositivo (s/pred estritamente-positivo?))
(def EstritamentePositivo (s/pred estritamente-positivo? 'estritamente-positivo?))

(pprint (s/validate EstritamentePositivo 15))
; (pprint (s/validate EstritamentePositivo 0))
; (pprint (s/validate EstritamentePositivo -15))

(println "\n versão 2")

(s/def Paciente
  "Schema de um paciente"
  {:id (s/constrained s/Int pos?), :nome s/Str})
;é por isso que é importante debulhar a documentação
; já existe pos? e já existe pos-int
; dica: sempre debulhar documentação


(pprint (s/validate Paciente {:id 15, :nome "guilherme"}))
;(pprint (s/validate Paciente {:id -15, :nome "guilherme"}))
;(pprint (s/validate Paciente {:id 0, :nome "guilherme"}))

; não é recomendado usar lambdas dentro dos schemas
;pois fica confuso ou perde a legibilidade do schema
;(s/def Paciente
;  "Schema de um paciente"
;  {:id (s/constrained s/Int #(> % 0)), :nome s/Str})












src/hospital_clojure/aula3.clj
(ns hospital-clojure.aula3
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

(def PosInt (s/pred pos-int?))

(def Paciente
  {:id (s/pred pos-int?), :nome s/Str})

(def Paciente
  {:id PosInt, :nome s/Str})

(s/defn novo-paciente :- Paciente
  [id :- PosInt
   nome :- s/Str]
  {:id id, :nome nome})

(pprint (novo-paciente 15, "Guilherme"))
;(pprint (novo-paciente -5, "Guilherme"))

(defn maior-ou-igual-a-zero? [x] (>= x 0))

(def ValorFinanceiro (s/constrained s/Num maior-ou-igual-a-zero?))

(def Pedido
  {:paciente     Paciente
   :valor        ValorFinanceiro
   :procedimento s/Keyword})

; Será que faz sentido "mini-schema" como aliases
; (def Procedimento s/Keyword)

(s/defn novo-pedido :- Pedido
  [paciente :- Paciente
   valor :- ValorFinanceiro
   procedimento :- s/Keyword]
  {:paciente paciente, :valor valor, :procedimento procedimento})

(pprint (novo-pedido (novo-paciente 15, "Amanda"), 15.33, :raio-x))
;(pprint (novo-pedido (novo-paciente 15, "Amanda"), -15.33, :raio-x))

(def Numeros [s/Num])

(pprint (s/validate Numeros [15]))
(pprint (s/validate Numeros [15, 13]))
(pprint (s/validate Numeros [15, 13, 132, 321, 23.2]))
(pprint (s/validate Numeros [0]))
;(pprint (s/validate Numeros [nil]))
(pprint (s/validate Numeros []))
(pprint (s/validate Numeros nil))

(def Plano [s/Keyword])
(pprint (s/validate Plano [:raio-x]))

(def Paciente
  {:id PosInt, :nome s/Str, :plano Plano})

(pprint (s/validate Paciente {:id 15, :nome "Guilherme", :plano [:raio-x, :ultrassom]}))

(pprint (s/validate Paciente {:id 15, :nome "Guilherme", :plano [:raio-x]}))
(pprint (s/validate Paciente {:id 15, :nome "Guilherme", :plano []}))
(pprint (s/validate Paciente {:id 15, :nome "Guilherme", :plano nil}))

;(pprint (s/validate Paciente {:id 15, :nome "Guilherme"}))

src/hospital_clojure/aula5.clj
(ns hospital-clojure.aula5
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(s/set-fn-validation! true)

(def PosInt (s/pred pos-int? 'inteiro-positivo))
(def Plano [s/Keyword])
(def Paciente
  {:id                          PosInt,
   :nome                        s/Str,
   :plano                       Plano,
   (s/optional-key :nascimento) s/Str})

(def Pacientes
  {PosInt Paciente})

(pprint (s/validate Pacientes {}))

(def Visitas
  {PosInt [s/Str]})

; romovido o IF e THROW pois
; o schema GARANTIU a existencia do ID e a Validade do ID
; ... se a validação estiver ativa
(s/defn adiciona-paciente
  [pacientes :- Pacientes, paciente :- Paciente]
  (let [id (:id paciente)]
    (assoc pacientes id paciente)))


(s/defn adiciona-visita :- Visitas
  [visitas :- Visitas, paciente :- PosInt, novas-visitas :- [s/Str]]
  (if (contains? visitas paciente)
    (update visitas paciente concat novas-visitas)
    (assoc visitas paciente novas-visitas)))


(s/defn imprime-relatorio-de-pacientes
  [visitas :- Visitas, paciente :- PosInt]
  (println "Visitas do Paciente" paciente "São:" (get visitas paciente)))

(defn testa-uso-de-paciente []
  (let [guilherme {:id 15, :nome "Guilherme", :plano []}
        daniela {:id 20, :nome "Daniela", :plano []}
        paulo {:id 25, :nome "Paulo", :plano []}

        ;uma variação com reduce, mais natural
        pacientes (reduce adiciona-paciente {} [guilherme, daniela, paulo])

        ;uma variação com shadowing, não fica bom
        visitas {}
        visitas (adiciona-visita visitas 15 ["01/01/2019"])
        visitas (adiciona-visita visitas 20 ["01/02/2019", "01/01/2020"])
        visitas (adiciona-visita visitas 15 ["01/03/2019"])]

    (pprint pacientes)
    (pprint visitas)

    ; grande problema pois estou usando o simbolo paciente
    ; em varios lucares do meu programa com significados diferentes
    (imprime-relatorio-de-pacientes visitas 20)))

(testa-uso-de-paciente)

src/hospital_clojure/aula1.clj
(ns hospital-clojure.aula1
  (:use clojure.pprint)
  (:require [schema.core :as s]))

(defn adicionar-paciente
  [pacientes, paciente]
  (if-let [id (:id paciente)]
          (assoc pacientes id paciente)
          (throw (ex-info "Paciente não possui id" {:paciente paciente}))))

(defn imprime-relatorio-de-pacientes
  [visitas, paciente]
  (println "Visitas do Paciente" paciente "São:" (get visitas paciente)))

; {15 [], 20 []}
(defn adiciona-visita
  [visitas, paciente, novas-visitas]
  (if (contains? visitas paciente)
    (update visitas paciente concat novas-visitas)
    (assoc visitas paciente novas-visitas)))

(defn testa-uso-de-paciente
  []
  (let [guilherme {:id 15, :nome "Guilherme"}
        daniela {:id 20, :nome "Daniela"}
        paulo {:id 25, :nome "Paulo"}
        ;uma variação com reduce, mais natural
        pacientes (reduce adicionar-paciente {} [guilherme, daniela, paulo])

        ;uma variação com shadowing, não fica bom
        visitas {}
        visitas (adiciona-visita visitas 15 ["01/01/2019"])
        visitas (adiciona-visita visitas 20 ["01/02/2019", "01/01/2020"])
        visitas (adiciona-visita visitas 15 ["01/03/2019"])]

    (pprint pacientes)
    (pprint visitas)

    ; grande problema pois estou usando o simbolo paciente
    ; em varios lucares do meu programa com significados diferentes
    (imprime-relatorio-de-pacientes visitas daniela)
    (println (get visitas 20))))

(testa-uso-de-paciente)

(pprint (s/validate long 15))
; (pprint (s/validate long "Amanda"))
; (pprint (s/validate long [15, 13]))

(s/set-fn-validation! true)

(s/defn teste-simples
  [x :- Long]
  (println x))

(teste-simples 30)
; (teste-simples "Amanda")

(s/defn imprime-relatorio-de-pacientes
  [visitas, paciente :- Long]
  (println "Visitas do paciente" paciente "São:" (get visitas paciente)))

; agora conseguimos o erro em tempo de *execução* que diz
; que o valor passado como parametro não condiz com o schema LongŁ
; (testa-uso-de-paciente)

(s/defn novo-paciente
  [id :- Long, nome :- s/Str]
  {:id id, :nome nome})

(println (novo-paciente 15 "Guilherme"))
; (println (novo-paciente "Guilherme" 15))
